Char bit instead of 8

TYuD opened the pull request.

pavel-kirienko reviewed (changes requested):
Thanks! A few initial changes are needed:

- [ ] Please remove the new XML, .pro, .ldf, and .bat files (we do not support Windows as a development platform; even if we decide to, PowerShell should be used instead).

- [ ] Please do not rely on memory aliasing, even in tests, if you can help it (things like `const AsInt* serialized = (const AsInt*)serialized_result;` are generally not acceptable).

- [ ] Add backward compatibility aliases for all of the `_BYTES_` constants that you renamed.

thirtytwobits commented:
> * Please remove the new XML, .pro, .ldf, and .bat files

You should be able to use cmake to generate

thirtytwobits commented:
sorry. Wrong button

thirtytwobits commented on src/nunavut/lang/c/support/serialization.j2:
```diff
@@ -161,26 +177,48 @@ static inline void nunavutCopyBits(void* const dst,
+        last_src  = psrc + length_chars;
+        last_dst  = pdst + length_chars;
+        mask      = (unsigned)((1U << length_mod) - 1U);
+        *last_dst = (*last_dst & (unsigned)(~mask)) | (*last_src & mask);
```
There are a ton of compiler warnings/errors for implicit sign conversions. This line generates just one of these. Please be sure you are compiling with all warnings-as-errors.

thirtytwobits commented on src/nunavut/lang/c/support/serialization.j2:
```diff
@@ -161,26 +177,48 @@ static inline void nunavutCopyBits(void* const dst,
+            const unsigned in      = ((unsigned)(psrc[src_off >> CHAR_SHIFT] >> src_mod) << dst_mod) & UCHAR_MAX;
+            const unsigned a       = pdst[dst_off >> CHAR_SHIFT] & (~mask);
+            const unsigned b       = in & mask;
+            pdst[dst_off >> CHAR_SHIFT] = a | b;
```
this assignment loses precision. Again, a lot of compiler warnings need to be cleaned up.

TYuD replied:
> Again, a lot of compiler warnings need to be cleaned up.

I use the default mingw settings. They're pretty strict. There are no warnings. I we'll have to  investigate

thirtytwobits replied:
I'm going to push an update to Nunavut HEAD in a bit that uses the latest dev container which has a very new version of clang I was building against. You'll be able to find these warning by using that. Regardless, once you start looking you can see a lot of signed/unsigned conversions going on and possible precision loss.

thirtytwobits replied:
See #297

thirtytwobits reviewed (changes requested):
Please rebase and cleanup compiler warnings. I'll be able to do more analysis once I have this building in our dev container.

TYuD commented:
> * Please remove the new XML, .pro, .ldf, and .bat

XML - accidentally added to the repository. Other files were added intentionally. *.pro and *.dpj project files perhaps can be replaced by cmake, but without *.ldf (linker script, memory map) executeble for TS-201 processor cannot be created. I must use it to test CHAR_BIT==32. What to do?

TYuD commented:
> * Please do not rely on memory aliasing, even in tests

In tests - no problem. There is another question. Do you remember our discussion of void* and char* pointers? My compiler warns when implicit casting unit_fast8_t*:
![image](https://user-images.githubusercontent.com/20935423/225433020-35ee3d50-6763-4e23-97ff-163f15611d1a.png)
Does it bother you? 
With void* pointer no problem. How do you feel about the opinion https://github.com/OpenCyphal/nunavut/issues/233#issuecomment-1465252357?
I think if we work with objects as with memory sections (copying, etc.) converting from void* to char* is safe and legal.

TYuD commented:
> * Add backward compatibility aliases for all of the `_BYTES_` constants that you renamed.

At first I wanted to leave _BYTE_ for backward compatibility. Then I removed it to make the problem clearer. The _BYTE_ name is incorrect. In C, a byte is a char (8, 16, 32 bits). Your _BYTE_ means exactly 8 bits, as spelled out in the DS DL specification.

Is it possible to leave _BYTE_, but in such a way that a 'deprecated' warning occurs when using it?

TYuD commented:
> You should be able to use cmake to generate

Hi, Scott! I still have to learn cmake

TYuD commented on src/nunavut/lang/c/__init__.py:
```diff
@@ -268,7 +268,7 @@ class _CFit(enum.Enum):
 
     def to_std_int(self, is_signed: bool) -> str:
-        return "{}int{}_t".format(("" if is_signed else "u"), self.value)
+        return "{}int_fast{}_t".format(("" if is_signed else "u"), self.value)
```
> we shouldn't change the Nunavut ABI contract

Did I understand correctly that the binary layout should be preserved for deserialized structures? Not just serialized.

If you use uint_leastXX_t instead of uint_fastXX_t, then on machines with CHAR_BIT==8 (the only ones you had before), the packaging of the fields will not change. Your customers will not be harmed. Switch to uint_leastXX_t?

thirtytwobits replied:
Right, for a given set of Nunavut configuration options and for the same DSDL inputs the type code generated should compile to be bit-for-bit the same as it is today. That is, the binaries that include ser/des only need to maintain interface and behavioural contracts but the structs must be bit-for-bit the same. You can test this easily by generating before and after with serialization disabled (will just generate PoCos). If we change from int to fast_int we are going to modify the alignment of some types on some platforms. This is not acceptable without a major revision to the Nunavut version number which we are not keen to do right now.

TYuD replied:
However, using uint_leastXX_t instead of uint_fastXX_t will not change the alignment of the structures. Is that not enough?

thirtytwobits replied:
Can we do something like
```
def to_std_int(self, is_signed: bool, optimization: typing.Optional[str]) -> str:
    if optimization is not None:
        if optimization == "fast":
...
        elif optimization == "small":
...
```

etc. Then you can provide these hints only where you have non-CHAR_BIT==8 code. Is that possible?

TYuD replied:
A lot of settings is also bad. Maybe it has sense to consider this possibility later.

In addition, all compilers, that I have deal with, define the types (u)int_fastXX_t the same as (u)int_leastXX_t. So, in most cases we don't lose anything, using (u)int_leastXX_t. Rather, the type (u)int_fastXX_t is provided for potential optimization

thirtytwobits commented:
> > * Please remove the new XML, .pro, .ldf, and .bat
> 
> XML - accidentally added to the repository. Other files were added intentionally. *.pro and *.dpj project files perhaps can be replaced by cmake, but without *.ldf (linker script, memory map) executeble for TS-201 processor cannot be created. I must use it to test CHAR_BIT==32. What to do?

I'm okay with having the linker script. I'd like to have it packaged up a bit differently so we can crosscompile on non-windows platforms but let's worry about the core PR first before we deal with that sort of rigging.

pavel-kirienko commented:
>In tests - no problem. There is another question. Do you remember our discussion of void* and char* pointers? My compiler warns when implicit casting unit_fast8_t*:

Aren't we supposed to use `unsigned char` instead of `unit*8_t` for maximum portability? For serialization purposes, we need the smallest integer type, which is guaranteed to be char.

>At first I wanted to leave BYTE for backward compatibility. Then I removed it to make the problem clearer. The BYTE name is incorrect. In C, a byte is a char (8, 16, 32 bits). Your >BYTE means exactly 8 bits, as spelled out in the DS DL specification.
>
>Is it possible to leave BYTE, but in such a way that a 'deprecated' warning occurs when using it?

I see what you're saying, but we don't want to break backward compatibility with existing code out there, so let us please keep `_BYTES_`. I don't think it's possible to mark a macro as deprecated, at least without relying on compiler-specific extensions, which we don't want to do (see misra). Also, this is Cyphal-related code, so our interpretation of "byte" as a a 8-bit number makes sense even if it is ambiguous in C.

TYuD commented:
> Aren't we supposed to use `unsigned char` instead of `unit*8_t` for maximum portability? For serialization purposes, we need the smallest integer type, which is guaranteed to be char.

Ok. Then we change the type of all buffers in all tests from uint8_t/uint_fast8_t to unsigned char? And all users of Cyphal will have to do the same. Yes?

> so let us please keep `_BYTES_`

Ok. Can I leave OCTETS in addition to BYTES? Two macros with the same value

TYuD commented:
> I don't think it's possible to mark a macro as deprecated

I like use enums:
`enum{ x_capacity = 100 };`
 instead of:
`#define X_CAPACITY 100` 
It obeys the scope of names etc. Can enums help to do deprecate? In C++ can, but in C...

pavel-kirienko commented:
>Ok. Then we change the type of all buffers in all tests from uint8_t/uint_fast8_t to unsigned char? And all users of Cyphal will have to do the same. Yes?

As I see it, since `uint8_t` and `unsigned char` are synonyms on all platforms where `uint8_t` is present, the changes will not affect the existing users. Do you agree?

>Ok. Can I leave OCTETS in addition to BYTES? Two macros with the same value

Okay.

pavel-kirienko commented:
>Can enums help to do deprecate? In C++ can, but in C...

Yes, but [only starting with C23](https://en.cppreference.com/w/c/language/attributes/deprecated).

Maybe we could replace the current `_BYTES_` macro with enums and define it roughly as follows:

```c++
#if deined(__cplusplus) || (defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 202311L))
[[deprecated]]
#endif
enum { FOO_BAR_BYTES_ = 123 };
```

TYuD commented:
> As I see it, since `uint8_t` and `unsigned char` are synonyms on all platforms where `uint8_t` is present, the changes will not affect the existing users. Do you agree?

In general, I agree, but some compilers issue a warnings. Are we going to ignore this?
![image](https://user-images.githubusercontent.com/20935423/226010598-666d8c54-f97d-4e18-93a7-f287638fc789.png)

pavel-kirienko commented:
>In general, I agree, but some compilers issue a warnings. Are we going to ignore this?

Per my understanding, one should be using `unsigned char` here (which is 32-bit wide) instead of `uint*32_t`.

TYuD commented:
> Per my understanding, one should be using `unsigned char` here (which is 32-bit wide) instead of `uint*32_t`.

Yes. If we abandon `void*`, then I recommend switching to unsigned char instead of `uint*_t` always. Including in the case of CHAR_BIT==8

Note also that the XXX_serialize function now accepts the buffer capacity not in octets, but in bits. This will affect the client code. It is reasonable to add the macro definition XXX_CAPACITY_BITS_

pavel-kirienko commented:
>Note also that the XXX_serialize function now accepts the buffer capacity not in octets, but in bits

I don't think this is wise as it breaks the API, which we really don't want to do. Can we go back to using either octets or bytes?

TYuD commented:
> I don't think this is wise as it breaks the API, which we really don't want to do. Can we go back to using either octets or bytes?

I understand you. And I don't want to change the interface either. Let's discuss the interface issue in more detail.

Here is the original interface (slightly simplified):

`int8_t X_serialize_(const X* obj, uint8_t* const buffer, size_t* const inout_buffer_size_bytes);`

And here's a new one:

`int X_serialize_(const X* obj, void* buffer, size_t buffer_size_bits, size_t* inout_offset_bits);`

Here are the following changes:

1) added an additional parameter `inout_offset_bits`. This is the offset from the beginning of the `buffer` before serialization (when entering the function) and the offset after serialization (when exiting the function);

2) `buffer_size_bits` is now only an input parameter (passed by value, not by pointer). It has only one meaning – the size of the entire `buffer`;

3) units are now bits, not octets.

The interface changes are caused by a fundamental problem. The fact is that DSDL objects are aligned by octets,but C by char. Let, for example, CHAR_BIT==32. Suppose we serialize a composite object and after serializing the first subobject, the offset become 3 octets. When serializing the second subobject, I cannot shift the char* pointer by 3*8=24 bits, but only by CHAR_BIT==32 bits. It follows that an additional parameter must be added to the function: the offset from the beginning of the buffer in fractions of CHAR_BIT.

In what units can the fractions of CHAR_BIT be expressed? In char, in octets, in bytes? These units are ambiguous and lead to confusion. The simplest thing is to express it in bits. Then the parameter will be unambiguous for any compilers.

Thus, the `inout_offset_bits` parameter appeared. It is important for us to set the offset before serialization, it is also important to know the offset after serialization. Therefore, we pass a pointer to `inout_offset_bits`. During serialization, we increment the contents of the variable to which this pointer points. The incremented value is available outside the serialization function. Therefore, `inout_offset_bits` is made input/output.

Having such the `inout_offset_bits pointer`, it becomes superfluous to pass the address of the `inout_buffer_size_bytes` parameter. It is enough to pass its value only.

What units should the buffer_size parameter have? The same problems with the ambiguity of char, octet and byte arise here as with the `inout_offset_bits` parameter. The confusion disappears if you take the bits. This I did.

There is another change with the void* pointer in the interface, but we have already discussed it enough.

So I tried to explain the essence of the problem. I hope you understand. In my opinion, interface changes are inevitable.

One of the compromise solutions seems as follow: write an implementation of `XXX_serialize_impl_` with a new interface. Make the old `XXX_serialize_` interface as a light wrapper over `XXX_serialize_impl_`. `Inside XXX_serialize_` one will recursively call `XXX_serialize_impl_`. If the user will serialize composite objects as a whole, and not individually by component parts, then this will be enough.

pavel-kirienko commented:
Thank you @TYuD for the explanation, it does make sense. Now I'm wondering if we should maybe aim for a major release of Nunavut v3 with these changes included so that we don't have to bother with backward compatibility? @thirtytwobits would that be acceptable?

pavel-kirienko commented:
I have no objections against 6-9 months.

TYuD commented:
> that we don't have to bother with backward compatibility?

The issue with backward compatibility seems to have been resolved. 

Here's what I got:

1) Returned the previous interface:

`int X_serialize_(const X* obj, unsigned char* buffer, size_t* inout_buffer_size_chars);`

This is a light wrapper over a new interface:

`int X_serialize_impl_(const X* obj, unsigned char* buffer, size_t buffer_size_bits, size_t* inout_offset_bits);`

The function `X_serialize_impl_` is called inside `X_serialize_` recursively for composite types

2) Returned the previous macro definitions: `X_BYTES_`

3) In deserialized structures, I replaced `(u)int_fastXX_t` with `(u)int_leastXX_t`. Now the binary alignment is preserved.

4) Returned pointers from `void*` to `unsigned char*`. Removed casting pointers from `void*` to `unsigned char*`. This is done in `X_serialize_`, `X_serialize_impl_`, `X_deserialize_`, `X_deserialize_impl_`, `nunavut\support\serialize.h`. Also in all unit tests unsigned char buffers returned.

To remove `void*`, I had to tinker with unit tests. Without `void*`, the user code becomes more bulky, contains a lot of unnecessary operations. This is bad, because unnecessary operations lead to potential errors.

When making changes, I noticed that your original functions `nunavutCopyBits` and `nunavutGetBits` (see `nunavut\support\serialize.h`) use explicit casting of pointers from `void*` to `uint8_t*`. So `MISRA` requirements are violated. In this regard, the requirement to remove `void*` pointers from the rest of the code seems incorrect.

Can I still roll back the changes and leave void*?

6) All unit tests were successful for `CHAR_BIT==8` and for `CHAR_BIT==32`

7) Checked only Little Endian. I don't understand yet what big-endian means for `CHAR_BIT==32`. Are the octets rearranged in it? Or are 32-bit portions rearranged in 64-bit words?

8) I can't force the compiler to issue warnings, as Scott asks. My 32-bit mingw5.30 does not issue any warnings with options –Wall, -Wextra

9) Tomorrow I will try to upload the corrected code to pull request

pavel-kirienko commented:
4. Please describe the required changes to the verification suite and why the usage of `unsigned char*` makes the user code more complicated.

8. `-Wall` does not actually enable all warnings. You need to enable the more specific ones manually, see https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html. Example:

https://github.com/OpenCyphal/nunavut/blob/712ee2c66a7a32bf408f7ccd4b262f6efbfe0dd3/verification/cmake/compiler_flag_sets/common.cmake#L21-L36

pavel-kirienko commented:
1. Rounding is a non-issue because the comparison takes care of that -- it is not exact.
2. This appears to be unavoidable as we want to support different `CHAR_BIT`. I see no problem there.
3. Entering same value in multiple places is expected of unit tests.

The code presented here is fine, and I see no issues with it.

----

>Here's what the code with void pointers would look like*:
>It is clearer and more compact, but the main thing is that there are no listed potential errors in it

The code is incorrect because it will break on a non-little-endian platform.

----

>It would be possible to improve the original example somewhat:

This example is likewise incorrect for the same reason.

----

Please do not change the original code beyond extending it to `CHAR_BIT != 0`.

----

>But inside the nunavutGetF64 function, a similar pointer casting is used.
>It seems that we are deceiving ourselves, for the sake of formal compliance with MISRA.

Compliance with MISRA is important but in any non-trivial projects deviations are unavoidable. This is explicitly recognized by MISRA itself as well. This does not mean that one should not strive to minimize the amount of deviations; further, the non-compliance you mentioned may need to be addressed as a matter of separate effort. We should also keep in mind that a localized non-compliance is significantly less costly than one exposed through a public API.

pavel-kirienko commented:
Closing as stale. Please re-submit to continue.
